### ISSUE TITLE
Interested in running static code analysis on Perl-Critic project?

### ISSUE THREAD

**Issue by otrosien** [Perl-Critic/Perl-Critic#717]
I was just chatting with @wbraswell about it, and he proposed creating a github issue, even though this is technically speaking not a Perl::Critic issue... I'm the author of sonar-perl [1], a perl plugin for SonarQube. On http://sonarqube.racodond.com/ I've been granted free hosting of a sonarqube instance running my perl plugin, and if you guys are interested I could provide an API key to put Perl-Critic there as well. In return I'd be interested in your feedback. Featues, as of now:

* LOC reporting (incl. comment and POD lines)
* Structural metrics (classes and functions)
* Test metrics (duration, successful and failed test count)
* Syntax highlighting of imported code
* Attaching of Perl::Critic issues
* Test Coverage (via Devel::Cover::Report::Clover)

Will has my email address - so if you want to, you can get in touch with me. A sample integration with travis-ci can be seen here: [2]

Oliver

[1] https://github.com/otrosien/sonar-perl
[2] https://github.com/otrosien/Perl-Critic-CognitiveComplexity

**Comment by wbraswell** [comment]
@petdance 
Do we have any reason not to want this?

**Comment by petdance** [comment]
It sounds useful, but I'd want to see some samples of what it does for other projects before we bother with @otrosien investing any time into it.

**Comment by wbraswell** [comment]
@petdance 
I think @otrosien meant to post this link, but forgot.  (He sent it to me on IRC #perlcritic.)

http://sonarqube.racodond.com/dashboard?id=Perl-Critic-CognitiveComplexity

**Comment by petdance** [comment]
Looks interesting.  Are there examples we can look at that do have bugs & code smells & (especially) duplications?

**Comment by otrosien** [comment]
I don't think I can recognize code duplications yet, but I created an issue for this (https://github.com/otrosien/sonar-perl/issues/42). Bugs/code smells are mainly populated by slurping in a perlcritic report. In addition to that, insufficient code coverage or insufficient documentation is defaulted to raise issues on your code. 

The main benefit comes when the project evolves, and you can watch these metrics change over time (like introducing more code smells, increasing/decreasing of code coverage, LOC metrics etc.)

I have some users running the plugin on a local SonarQube, but the public hosting offered on racodond.com is still fairly new, so there is no major perl project actively using it yet. The "mojolicous" project over there is just a test run I created, and will be removed again.

In a nutshell, I'd be happy to support you, and at the same time gain some feedback from you.

**Comment by petdance** [comment]
If you want to go ahead and set it up to check Perl::Critic, I'd be glad to see what comes out of it.  I can't guarantee that we'll use it, but I'd certainly be glad to look at it.  It might be incredibly beneficial, or we might ignore it completely.  At this point I can't say either way, but I am definitely curious as to what it might turn up.

Thanks for offering to help.

**Comment by otrosien** [comment]
Sure, I'll go and issue a PR. How should the .perlcriticrc look like, that you want the scanner to use?

**Comment by petdance** [comment]
No idea.  Can you just come up with one for this initial version?  Then we can tweak from there.

**Comment by wbraswell** [comment]
To start with, I suggest:
--gentle --theme=core

**Comment by otrosien** [comment]
done.

**Comment by petdance** [comment]
Committed and running.  See #721 for tweaks we need.

**Comment by drmmr763** [comment]
I know this has been closed for a time but I'm intending on using @otrosien's sonarqube perl code for my own project shortly. In researching setting that up I ran across https://about.sonarcloud.io/ which offers free FOSS project hosting for SonarQube projects. Seems like that could be a new home for the Perl Critic project?

**Comment by petdance** [comment]
That would be great.  If you get that going, please open a new ticket and refer back to this one.

**Comment by otrosien** [comment]
I haven't had time to find new reliable hosting for the perl sonarqube plugin yet. I doubt this will be available as free service anywhere, but I haven't looked much into potentially free offerings for open source projects on platforms like heroku etc.

With regards to sonarcloud - they only host the official sonarqube plugins, not third-parties. This was the reason to host it at racodond.com which was a site run by a fellow sonarqube plugin developer who unfortunately resigned. But I'm running the code analysis for the sonar-perl plugin project itself over there (it's java-based): https://sonarcloud.io/dashboard?id=com.github.otrosien%3Asonar-perl
